File: src/main/services/diagnostics/DiagnosticsService.ts
import { randomUUID } from 'node:crypto'
import {
  DiagnosticsSnapshotSchema,
  DiagnosticLogEntrySchema,
  type BuildInfo,
  type DiagnosticLevel,
  type DiagnosticLogEntry,
  type DiagnosticsSnapshot
} from '@shared/contracts'

type Publisher = (channel: string, payload: unknown) => void

export class DiagnosticsService {
  private readonly entries: DiagnosticLogEntry[] = []
  private readonly build: BuildInfo
  private publisher?: Publisher

  constructor(build: BuildInfo) {
    this.build = build
  }

  setPublisher(publisher: Publisher): void {
    this.publisher = publisher
  }

  getSnapshot(): DiagnosticsSnapshot {
    return DiagnosticsSnapshotSchema.parse({
      build: this.build,
      entries: this.entries
    })
  }

  getEntries(): DiagnosticLogEntry[] {
    return [...this.entries]
  }

  clear(): void {
    this.entries.splice(0, this.entries.length)
  }

  log(level: DiagnosticLevel, source: string, message: string, detail?: string): DiagnosticLogEntry {
    const entry = DiagnosticLogEntrySchema.parse({
      id: randomUUID(),
      timestamp: new Date().toISOString(),
      level,
      source,
      message,
      detail
    })

    this.entries.unshift(entry)
    if (this.entries.length > 300) {
      this.entries.length = 300
    }

    this.publisher?.('diagnostics:event', entry)
    return entry
  }
}

File: WsaManager.Core/DiagnosticsService.cs
namespace WsaManager.Core;

public sealed class DiagnosticsService
{
    private readonly List<DiagnosticLogEntry> entries = [];
    private readonly object syncRoot = new();

    public event EventHandler<DiagnosticLogEntry>? EntryAdded;

    public IReadOnlyList<DiagnosticLogEntry> Entries
    {
        get
        {
            lock (syncRoot)
            {
                return entries.ToList();
            }
        }
    }

    public DiagnosticLogEntry Log(DiagnosticLevel level, string source, string message, string? detail = null)
    {
        var entry = new DiagnosticLogEntry
        {
            Id = Guid.NewGuid().ToString("n"),
            Timestamp = DateTimeOffset.UtcNow,
            Level = level,
            Source = source,
            Message = message,
            Detail = detail
        };

        lock (syncRoot)
        {
            entries.Insert(0, entry);
            if (entries.Count > 300)
            {
                entries.RemoveRange(300, entries.Count - 300);
            }
        }

        EntryAdded?.Invoke(this, entry);
        return entry;
    }

    public void Clear()
    {
        lock (syncRoot)
        {
            entries.Clear();
        }
    }
}
